nexus/db-queries/src/db/datastore/physical_disk.rs

```diff
@@ -126,6 +118,82 @@ impl DataStore {
         Ok(())
     }
 
+    /// Returns all physical disks which:
+    ///
+    /// - Appear on in-service sleds
+    /// - Appear in inventory
+    /// - Do not have any records of expungement
+    pub async fn physical_disk_uninitialized_list(
+        &self,
+        opctx: &OpContext,
+        inventory_collection_id: Uuid,
+        sled_id: Uuid,
+    ) -> ListResultVec<InvPhysicalDisk> {
+        opctx.authorize(authz::Action::Read, &authz::FLEET).await?;
+
+        use db::schema::inv_physical_disk::dsl as inv_physical_disk_dsl;
+        use db::schema::physical_disk::dsl as physical_disk_dsl;
+        use db::schema::sled::dsl as sled_dsl;
+
+        // This is kind of frustrating - we'd like to be able to apply this
+        // condition to all "in service" sleds, regardless of provisionability,
+        // but the sled policy crate explicitly avoids exporting that type.
+        //
+        // As a workaround, we must create the corresponding external types
+        // and manually convert them back into the database types.
+        let in_service = to_db_sled_policy(SledPolicy::InService {
+            provision_policy: SledProvisionPolicy::Provisionable,
+        });
+        let in_service_non_provisionable =
+            to_db_sled_policy(SledPolicy::InService {
+                provision_policy: SledProvisionPolicy::NonProvisionable,
+            });
+
+        sled_dsl::sled
+            // If the sled is not in-service, drop the list immediately.
+            .filter(
+                sled_dsl::sled_policy
+                    .eq(in_service)
+                    .or(sled_dsl::sled_policy.eq(in_service_non_provisionable)),
+            )
+            // Look up all inventory physical disks that could match this sled
+            .inner_join(
+                inv_physical_disk_dsl::inv_physical_disk.on(
+                    inv_physical_disk_dsl::inv_collection_id
+                        .eq(inventory_collection_id)
+                        .and(inv_physical_disk_dsl::sled_id.eq(sled_id))
+                        .and(
+                            inv_physical_disk_dsl::variant
+                                .eq(PhysicalDiskKind::U2),
+                        ),
+                ),
+            )
+            // Filter out any disks in the inventory for which we have ever had
+            // a control plane disk.
```

Comment:
This is probably a naive question because I don't grok the disk lifecycle yet, but: does this prevent us from reinitializing a disk that has previously been expunged?

Reply:
Yes, I believe that is intentional based on last nights discussion, and possibly RFD 457.